client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java

```diff
@@ -0,0 +1,157 @@
+/*******************************************************************************
+ * Copyright (c) 2019 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.client.impl;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import org.apache.qpid.proton.amqp.messaging.Modified;
+import org.apache.qpid.proton.amqp.messaging.Rejected;
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.client.Command;
+import org.eclipse.hono.client.CommandContext;
+import org.eclipse.hono.client.DelegatedCommandSender;
+import org.eclipse.hono.client.MessageSender;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.MessageHelper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.proton.ProtonDelivery;
+
+/**
+ * Handler for command messages that delegates command handling either to a given command handler or
+ * to a matching consumer via the downstream peer.
+ */
+public class DelegatingCommandHandler implements Handler<CommandContext> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DelegatingCommandHandler.class);
+
+    private final Function<String, Handler<CommandContext>> deviceSpecificCommandHandlerSupplier;
+    private final Function<String, Future<DelegatedCommandSender>> delegatedCommandSenderSupplier;
+
+    /**
+     * Creates a new DelegatingCommandHandler.
+     *
+     * @param deviceSpecificCommandHandlerSupplier Function to get an existing command handler for a given device.
+     * @param delegatedCommandSenderSupplier Supplies a sender for a given tenant to send the delegated command via the
```

Comment:
so, the function's parameter is the tenant ID?
can the returned Future fail? what if there is no sender for the given tenant?

Reply:
I've adapted the javadoc.